Update of CONTRIBUTING file (#2805)
cntt-n/CNTT: CONTRIBUTING.md

  - **Read** will be assigned to those who want to create issues, join the discussions, receive notifications, and provide feedback.
  - **Triage** will be assigned to active contributors who create issues, pull requests and review them.
  - **Writes** will be assigned to sub-project Co-Leads to allow them to manage issues and pull request, approve them and be auto requested for review.
  - **Maintain** will be assigned to sub-project Leads to allow them to merge PRs into the repository.

<a name="labels"></a>
### Labels

Every Issue and PR should be tagged with one or more of the following labels to indicate their status and for the automation pipelines to apply the correct Anuket processes.

  - **Backlog**: The Issue will be addressed in future releases..
  - **Major/Minor**: The Issue or PR concerns a major or minor change.. 
  - **WIP**: The PR is still under development, no reviews/approvals are expected yet.

>      - PRs can also be created as "Draft" to indicate that the PR is in progress.

Reviewer: what does it mean "PR is in progress"? 
I think the distinction between WIP and Draft needs to be clearer. For example, WIP may indicate that the content is incomplete but what exists is correct, whereas Draft would indicate that the content is not organized, has not been verified, and may contain errors.

Author: WIP and Draft are synonyms in this context. Should I make this clear?

Reviewer: @CsatariGergely Or just delete one of them including from the "Labels" list. Thanks

Author: Corrected in [c8c328e](https://github.com/cntt-n/CNTT/pull/2805/commits/c8c328ea479ffbf74fb060e8e8e7ad8630a0f4d2)

* [Idle Pull Requests Policy](#idle)
* [Unresolved Pull Requests Policy](#unresolved)

> This document is an extension to the generic [Anuket Project Operations and Procedures](https://wiki.anuket.io/display/HOME/Anuket+Project+Operations+and+Procedures) and describes contribtion rules to the Anuket specifications, that is the set of documents maintained in this repository. 

Reviewer 2: ```suggestion
This document is an extension to the generic [Anuket Project Operations and Procedures](https://wiki.anuket.io/display/HOME/Anuket+Project+Operations+and+Procedures) and describes contribution rules to the Anuket specifications, that are the set of documents maintained in this repository. 
```

* [Idle Pull Requests Policy](#idle)
* [Unresolved Pull Requests Policy](#unresolved)

This document is an extension to the generic [Anuket Project Operations and Procedures](https://wiki.anuket.io/display/HOME/Anuket+Project+Operations+and+Procedures) and describes contribution rules to the Anuket specifications, that are the set of documents maintained in this repository. 

<a name="permissions"></a>

> ### Permission levels

Reviewer 3: Workstream to be changed by sub-project in this section

Author: Workstreams corrected in [fc58bc5](https://github.com/cntt-n/CNTT/pull/2805/commits/fc58bc510b0c6520b48196696396b8c32f94d641)

There are different permissions (levels) available for contributing into the Anuket specifications:
- **Outside Collaborators**: This includes general public. 
  - Outside collaborators can create issues and add comments to issues and Pull Requests.

> - **Read, Triage, Write, and Maintain** Permissions, which are explained in [here](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization#permission-levels-for-repositories-owned-by-an-organization), will be assigned as follows:

Reviewer 4: ```suggestion
- **Read** will be assigned to those who want to create issues, join the discussion, receive notifications, and provide feedback.

Reviewer 4: This was put in the wrong place.  It refers to the line below that needs to change.

Reviewer: @bfcohen100 something is missing in your line above - "read" doesn't allow rights to create

Author: I think it allows the creation of issues.

Author: @bfcohen100 I do not see this misplaced. Line 20 lists all the non outside collaborators and lines 21-24 explain each of the,

<a name="issues"></a>
### Issues

- Issues must be correctly tagged with the correct labels and Milestone by the sub-project lead.
  - Title needs to be in the following format [SPx Chx] (if applicable).
    - SPx: The name of the sub-project it is targeting.
    - Chx: The name of the chapter it is targeting (if applicable).
- Upon Creation, Issues must be correctly set with the correct **Project Name**. 
  - Project Name = Sub-project name.

> - Issues must be correctly set with the right labels and Milestone by the sub-project lead.

Author: Corrected in [fd79e24](https://github.com/cntt-n/CNTT/pull/2805/commits/fd79e24066d856d1885d48b991f2660f24f7949e)

  - Relevant Milestone for the upcoming release (M1 - M6).
- Contributors can assign an unassigned issue to themselves if they would like to work on them
- Sub-project leads have the responsibility to assign stalling or priority issues to individuals to address them.

> - Sub-project lead may label an issue as "**Returned**" to request further information or to decide not to work on it.

Reviewer 2: ```suggestion
- Sub-project leads may label an issue as "**Returned**" to request further information or to decide not to work on it.
```

- Only the person assigned ("assignee") to a PR may edit it. The normal process is for others to make suggests in "Comments" but not directly edit a PR unless the assignee has granter another contributor permission to contribute directly to the PR.
- A PR is recommended to only create/modify content within the scope of a single chapter in a single sub-project if possible.
- A PR should only add/change content related to the issue associated with the PR.
- Comments/sub-Conversations within a PR may only be marked Resolved by:
  - A) The originator of the sub-conversation. 
  - B) The relevant sub-project lead or 
  - C) An automated timeout of 1 week when there has been no additional comments after an update is posted to resolve the sub-thread. 
  - **The person responding to a sub-conversation shall not mark it as Resolved**.
- All changes to an existing PRs shall be made through subsequent commits. 
  - **Do not modify an existing commit, create separate commits under the same PR**.
  - After major changes to a PR reapproval is needed from all previous approvers
  - PRs will be merged using "Squash and Merge" to enforce a linear commit history.

> 

Reviewer: can we add that the person performing the merge shall ensure that the branch is deleted

Author: to be honest I would like to add a bullet point saying that PR-s should be created from forks instead from the upstream repo. In this way the upstream repo would not be polluted with development branches.

Reviewer 4: Agree with Gergely here.

Reviewer: because of permission issues at AT&T I was never able to fork. I haven't tried at MSFT as am waiting for GitLab conversion.

Author: > because of permission issues at AT&T I was never able to fork. I haven't tried at MSFT as am waiting for GitLab conversion.

Maybe we can discuss this on the TSC. If it is not mandatory for you to open pr-s from MSFT-s GitHub org you can simply use your personal GitHub org.

Author: > Also, there are minor updates that are made during some meetings and approved in it.

Sorry, I do not understand this comment. Can you please explain?

Reviewer: @CsatariGergely This is with respect to forks -- what I am pointing out is that in many sub-project meetings we have made a quick change (typo, grammar, etc.) and approved it all during the same meeting. Creating a fork etc. would be a burden in such circumstances.

<a name="approvals"></a>
### Pull Requests Approval Process

- Once a PR is created, it needs to get the following approvals before it is merged into master.

>   - From at least one of contributor for that sub-project

Reviewer: should this be at least 3 also "contributor" should be plural

Author: Corrected in [fd79e24](https://github.com/cntt-n/CNTT/pull/2805/commits/fd79e24066d856d1885d48b991f2660f24f7949e)

    - Approvers should be distributed equally among operators and vendors.
  - From sub-project lead (to make sure consensus is reached).
  - Sub-project lead can request additional approvals reviews.
    - Recommendation is to use 4 approvals for complex PRs.
  - 2 business days cool off period should be applied before Final approval.

>   - Only One approval will be counted per each organization for a given PR.

Reviewer 2: ```suggestion
  - Only one approval will be counted per each organization for a given PR.
```

- The selection of which contributor to approve a PR is made by the sub-project lead and should take those factors into consideration:
  - Contributor needs to be actively discussing the PR to be selected for approval.
- PRs will be merged automatically online by the sub-project lead or other committers of the sub-projetc once consensus is reached and all approvals are received. 

>   - Sub-projects shold have an agreed set of committers who are trusted with merging rights to the sub-projects. [CODEOWNERS](CODEOWNERS) file should contain all the committers of the sub-projects including the sub-proejct lead.

Reviewer 5: ```suggestion
  - Sub-projects should have an agreed set of committers who are trusted with merging rights to the sub-projects. [CODEOWNERS](CODEOWNERS) file should contain all the committers of the sub-projects including the sub-project lead.
```

Reviewer 4: ```suggestion
  - Sub-projects should have an agreed set of committers who are trusted with merging rights to the sub-projects. [CODEOWNERS](CODEOWNERS) file should contain all the committers of the sub-projects including the sub-project lead.
```

Reviewer: @bfcohen100 I am missing the difference from what @gkunz suggested

> - If sub-project lead is the person who is creating the PR, They need to request an alternate approver, preferably from a co-lead or from the contributors list.

Reviewer 5: ```suggestion
- If sub-project lead is the person who is creating the PR, they need to request an alternate approver, preferably from a co-lead or from the contributors list.
```

Reviewer: Can we generalise:
"The person creating the PR cannot be an approver and should not be the person performing the merge."

Reviewer 5: I like it and I modified the suggested change accordingly.

> - If sub-project lead is the person who is creating the PR, They need to request an alternate approver, preferably from the co-lead or from the contributors list.

Reviewer 2: ```suggestion
- If sub-project lead is the person who is creating the PR, They need to request an alternate approver, preferably from a co-lead or from the contributors list.
```

I think there could be cases with more than 2 leads, or at least this doesn't "shut that door."

> - If a PR affects more than one sub-projects or areas ourside of the scope of the sub-projects the following approvals are needed:

Reviewer 5: ```suggestion
- If a PR affects more than one sub-project or areas ourside of the scope of the sub-projects the following approvals are needed:
```

Reviewer: Typo: "... areas outside of the scope of the sub-project, the "

Reviewer 5: thanks, I updated the suggestion

>   - From the sub-project leads of the affected sub-projects or if the PR affects no sub-projects at least 3 contributors of Anuket specifications

Reviewer: maybe the 3 approving contributors need to be from different sub-projects?

Author: Corrected in [fd79e24](https://github.com/cntt-n/CNTT/pull/2805/commits/fd79e24066d856d1885d48b991f2660f24f7949e)

- A PR is recommended to only create/modify content within the scope of a single chapter in a single sub-project if possible.
- A PR should only add/change content related to the issue associated with the PR.
- Comments/sub-Conversations within a PR may only be marked Resolved by:
  - A) The originator of the sub-conversation. 
  - B) The relevant sub-project lead or 
  - C) An automated timeout of 1 week when there has been no additional comments after an update is posted to resolve the sub-thread. 
  - **The person responding to a sub-conversation shall not mark it as Resolved**.
- All changes to an existing PRs shall be made through subsequent commits. 
  - **Do not modify an existing commit, create separate commits under the same PR**.
  - After major changes to a PR reapproval is needed from all previous approvers
  - PRs will be merged using "Squash and Merge" to enforce a linear commit history.

<a name="approvals"></a>
### Pull Requests Approval Process

- Once a PR is created, it needs to get the following approvals before it is merged into master.
  - From at least three of contributors for that sub-project

>     - Approvers should be distributed equally among operators and vendors.

Reviewer 4: ```suggestion
    - Approvers should be distributed equally among operators and vendors.
```Remove this line or reword to make it more of a best practice.  This is going to vary depending on the doc/project

  - From sub-project lead (to make sure consensus is reached).
  - Sub-project lead can request additional approvals reviews.
    - Recommendation is to use 4 approvals for complex PRs.

>   - 2 business days cool off period should be applied before Final approval.

Reviewer 4: ```suggestion
  - 2 business days waiting period should be applied before Final approval.
```Do we really need this?

Reviewer 3: For me, the 2 business days waiting is not necessary

Author: 2 days is a bit too much, but I think 1 day waiting period we need.

>   - Only one approval will be counted per each organization for a given PR.

Reviewer 4: This might be hard to enforce in practice.
